torch/nn/utils/parametrize.py

```diff
@@ -655,3 +655,19 @@ def type_before_parametrizations(module: Module) -> type:
         return module.__class__.__bases__[0]
     else:
         return type(module)
+
+def transfer_parametrizations_and_params(from_module: Module, to_module: Module) -> Module:
+    r"""Transfers parametrizations and the parameters they parametrize from from_module
+    to to_module.
+
+     Args:
+        from_module (nn.Module): module to transfer from
+        to_module (nn.Module): module to transfer to
+    """
+    if is_parametrized(from_module):
+        assert isinstance(from_module.parametrizations, ModuleDict)
+        for parameter_name in from_module.parametrizations:
+            for param_func in from_module.parametrizations[parameter_name]:
+                setattr(to_module, parameter_name, from_module.parametrizations[parameter_name].original)
```

Comment:
I agree. This should be done using the current API. I think this, as implemented, should break, at least, caching.

Reply:
@Lezcano @albanD can you help me understand the issue, this function isn't removing any parametrizations, its just duplicating them onto a new module

Reply:
The caching system registers the module id and the tensor id into a cache and remembers the result.

https://github.com/pytorch/pytorch/blob/a8e45b596910f90013637ccc735d6ca21a93d852/torch/nn/utils/parametrize.py#L321

It's not clear to me whether this would work after calling `transfer_parametrizations_and_params`.

Furthermore, I don't think this is really correct. For once, you are registering the parametrizations from the outside to the inside, which is the inverse order they were added in the `from_module`.

Could you add some tests for this function that tests it when registering more than one parametrisation and test the caching as well?

I don't understand the `setattr` either, as the parameters within the parametrisation would also be reachable within the `to_module` by going through `to_module.parametrizations[parameter_name].original`. The tests should also test that this is not the case.

Reply:
OK, I have added the tests and verified that it doesn't break caching and that the ordering seems to be correct. I am not understanding the last paragraph of your comment though @Lezcano.

> don't understand the setattr either, as the parameters within the parametrisation would also be reachable within the to_module by going through to_module.parametrizations[parameter_name].original. The tests should also test that this is not the case.

I'm unsure why that is odd and/or needs to be tested. If you use any other method to transfer a weight or something from one module to another, isn't that how you would expect it to behave?

i.e. if you do something like

```
lin = torch.nn.Linear(in, out)
new_lin = torch.nn.qat.Linear(in, out, qconfig)
new_lin.weight = lin.weight
```

My goal was to do that, but transfer the value and the parametrizations as well.

I updated the transfer function a bit to match this intent.

Reply:
Great to know that everything's working as expected! I'll go over the tests later today. Sorry for the hassle :)

My point here is that, what you're doing at the moment is to set `other.weight = original`, and then you're registering parametrisations on top. Registering these parametrizations, if they have `right_inverse`s, would change that weight. What you should do is to set the attr to `getattr(from_module, parameter_name)` I believe. Now, if `not is_parametrized(module, parameter_name)`, you can also do it the other way around: You first register the parametrization, and then do `to_module.parametrizations[parameter_name].original = from_module.parametrizations[parameter_name].original`, or something of sorts. This latter approach would be preferred whenever possible for numerical stability reasons.

Reply:
OK I will investigate, is there a better test example I can use to verify that I'm doing it correctly once I fix this or should I just look for any transform with a right inverse?

Reply:
Already using one parametrisation with a `right_inverse` should be enough to show the problem. If that test passes I think that this should be ready to go.

Reply:
oh yeah, thanks for identifying this, that bug existed and on top of that a worse bug existed where the original module would get changed when transferring the parametrization